src/mesh.cpp

```diff
@@ -1330,6 +882,488 @@ void RectilinearMesh::to_hdf5(hid_t group) const
+  write_dataset(mesh_group, "p_grid", grid_[1]);
+  write_dataset(mesh_group, "z_grid", grid_[2]);
+
+  close_group(mesh_group);
+}
+
+//==============================================================================
+// SphericalMesh implementation
+//==============================================================================
+
+SphericalMesh::SphericalMesh(pugi::xml_node node) : StructuredMesh {node}
+{
+  n_dimension_ = 3;
+
+  grid_[0] = get_node_array<double>(node, "r_grid");
+  grid_[1] = get_node_array<double>(node, "t_grid");
+  grid_[2] = get_node_array<double>(node, "p_grid");
+
+  if (int err = set_grid()) {
+    fatal_error(openmc_err_msg);
+  }
+}
+
+const std::string SphericalMesh::mesh_type = "spherical";
+
+std::string SphericalMesh::get_mesh_type() const {
+  return mesh_type;
+}
+
+StructuredMesh::MeshIndex SphericalMesh::get_indices(Position r, bool& in_mesh) const
+{
+  Position mapped_r;
+
+  mapped_r[0] = r.norm();
+  if (mapped_r[0] < FP_PRECISION) {
+    mapped_r[1] = 0.0;
+    mapped_r[2] = 0.0;
+  } else {
+    mapped_r[1] = std::acos(r.z / mapped_r.x);
+    mapped_r[2] = std::atan2(r.y, r.x);
+    if (mapped_r[2] < 0) mapped_r[2] += 2*M_PI;
+  }
+   
+  MeshIndex idx = StructuredMesh::get_indices(mapped_r, in_mesh);
+
+  mapped_r[1] = sanitize_theta(mapped_r[1]);
+  mapped_r[2] = sanitize_phi(mapped_r[2]);
+
+  return idx;
+}
+
+double SphericalMesh::find_r_crossing(const Position& r, const Direction& u, double l, int shell) const 
+{
+  
+  if ((shell < 0) || (shell >= shape_[0])) return INFTY;
+  
+  // solve |r+s*u| = r0
+  // |r+s*u| = |r| + 2*s*r*u + s^2 (|u|==1 !)
+
+  const double r0 = grid_[0][shell];  
+  const double p = r.dot(u);    
+  double D = p * p - r.dot(r) + r0 * r0;
+  
+  if (D >= 0.0) {
+    D = std::sqrt(D);
+    // the solution -p - D is always smaller as -p + D : Check this one first
+    if (-p - D > l)
+      return -p - D;
+    if ( -p + D > l)
+      return -p + D;
+  }
+
+  return INFTY;
+}
+
+double SphericalMesh::find_theta_crossing(const Position& r, const Direction& u, double l, int shell) const 
+{
+  // Theta grid is [0, 180], thus there is no real surface to cross
+  if (full_theta and (shape_[1]==1)) return INFTY;
+  
+  shell = sanitize_theta(shell);
+  
+
+  // solving z(s) = cos/theta) * r(s) with r(s) = r+s*u
+  // yields 
+  // a*s^2 + 2*b*s + c == 0 with 
+  // a = cos(theta)^2 - u.z * u.z
+  // b = r*u * cos(theta)^2 - u.z * r.z
+  // c = r*r * cos(theta)^2 - r.z^2
+  
+  const double cos_t  = std::cos(grid_[1][shell]);
+  const bool sgn = std::signbit(cos_t);
+  const double cos_t_2 = cos_t * cos_t;
+
+  const double a = cos_t_2 - u.z * u.z;
+  const double b = (r.dot(u) * cos_t_2 - r.z * u.z);
+  const double c = (r.dot(r) * cos_t_2 - r.z * r.z);
+
+  // if factor of s^2 is zero, direction of flight is parallel to theta surface
+  if (fabs(a) < FP_PRECISION) {
+    // if b vanishes, direction of flight is within theta surface and crossing is not possible
```

Comment:
Based on the comment here, I'd make the opposite case an early exit for readability:

```cpp
    // if b vanishes, direction of flight is within theta surface and crossing is not possible
   if (fabs(b) >= FP_PRECISION) return INFTY;

   const double s = -0.5 * c / b;
   if (s > l) and (std::signbit(r.z+s*u.z) == sgn)) return s;
```